File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 1:
            return 1
        counter = dict()
        maxLen = 0
        left = 0
        right = 0
        currLen = 0
        while right < len(s):
            currChar = s[right]
            if currChar in counter.keys():
                counter[currChar] += 1
                currLen += 1
                while counter[currChar] > 1:
                    counter[s[left]] -= 1
                    left += 1
                    currLen -= 1
            else:
                counter[currChar] = 1
                currLen += 1
            right += 1
            maxLen = max(maxLen, currLen)
        return maxLen
                    

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> lastPos = new HashMap<>();
        int ans = 0;
        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            Character c = s.charAt(right);
            if (!lastPos.containsKey(c)) {
                lastPos.put(c, right);
                ans = Math.max(ans, right - left + 1);
            } else {
                if (lastPos.get(c) < left) {
                    lastPos.put(c, right);
                    ans = Math.max(ans, right - left + 1);
                } else {
                    left = lastPos.get(c) + 1;
                    ans = Math.max(ans, right - left + 1);
                    lastPos.put(c, right);
                }
            }
        }
        return ans;
    }
}